src/Core/src/Credentials/PsrCacheProvider.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
+
+namespace AsyncAws\Core\Credentials;
+
+use AsyncAws\Core\Configuration;
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * Cache the Credential generated by the decorated CredentialProvider with PSR-6.
+ *
+ * The Credential will be reused until it expires.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+final class PsrCacheProvider implements CredentialProvider
+{
+    private $cache;
+
+    private $decorated;
+
+    public function __construct(CredentialProvider $decorated, CacheItemPoolInterface $cache)
+    {
+        $this->decorated = $decorated;
+        $this->cache = $cache;
+    }
+
+    public function getCredentials(Configuration $configuration): ?Credentials
+    {
+        $item = $this->cache->getItem('AsyncAws.Credentials.' . sha1(\serialize([$configuration, \get_class($this->decorated)])));
+        if (!$item->isHit()) {
+            $item->set($credential = $this->decorated->getCredentials($configuration));
+
+            if (null !== $credential && $credential->isCacheable()) {
+                $item->expiresAt($credential->getExpireDate());
```

Comment:
What if `$credential->getExpireDate()` returns null?

Reply:
`$item->expiresAt(null)` is valid and use the default pool TTL

Reply:
But if the default cache for the pool is 10 years, then this will break the application eventually.

Reply:
but if the credential does not expires, then why couldn't we cache it for 10 years?

Note: All credentials fetched from HTTP have an expireDate. Maybe we should cache only credential with expiration?

Reply:
That is definitely more safe.

Reply:
done. And it also removes the `cacheable` propery in credential

Reply:
thanks for the hint